### PR TITLE
New Enterprise Module : JAAS Security

### PR DESCRIPTION
**Pull request by dbrimley**
This takes the JAAS Security example in my personal repo.  AFAIK, we do not currently have a JAAS Security example in the code samples project. 

Initial commit
Java classes
Fixed license issues
Change URL in README.md for obtaining license
ignore .java-version from jenv

### CODE REVIEW

File: enterprise/jaas-security/src/main/java/Member.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        hz.getClientService().addClientListener(new ClientListener() {
+
+            public void clientConnected(Client client) {
+                ClientEndpoint clientEndpoint = (ClientEndpoint) client;
+                System.out.println("Connected");
+            }
+
+            public void clientDisconnected(Client client) {
+
```
Comment (mesutcelik) [inline enterprise/jaas-security/src/main/java/Member.java]: can be added some logging..

File: enterprise/jaas-security/src/main/resources/hazelcast.xml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                <interface>127.0.0.1</interface>
+            </tcp-ip>
+        </join>
+    </network>
+
+    <security enabled="true">
+        <client-login-modules>
+            <login-module class-name="ClientLoginModule" usage="REQUIRED">
+                <properties>
+                    <property name="lookupFilePath">value3</property>
```
Comment (mesutcelik) [inline enterprise/jaas-security/src/main/resources/hazelcast.xml]: What is this property for?

File: enterprise/jaas-security/src/main/java/ClientLoginModule.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+/**
+ * An example ClientLoginModule that hard codes authorisation details in a pair of Maps.  You could amend this class to
+ * perform look up against an LDAP store that would then return a set of Groups for the User.
+ * <p>
+ * Obviously you would NEVER store passwords in clear text.
+ *
+ */
+public class ClientLoginModule implements LoginModule  {
+
```
Comment (mesutcelik) [inline enterprise/jaas-security/src/main/java/ClientLoginModule.java]: I suggest this impl to be reviewed by @tkountis @kwart

**Comment by mesutcelik** [comment]
@dbrimley Thanks for the sample! Can you please fix checkstyle problems?

File: enterprise/jaas-security/src/main/resources/hazelcast.xml (hunk 2 above)
Comment (dbrimley) [reply enterprise/jaas-security/src/main/resources/hazelcast.xml]: It's a way for the JAAS ClientLoginModule to take outside properties.  I guess this isn't very clear.  I'll change the naming.

File: enterprise/jaas-security/src/main/java/Member.java (hunk 1 above)
Comment (dbrimley) [reply enterprise/jaas-security/src/main/java/Member.java]: ok, added.

File: enterprise/jaas-security/src/main/java/ClientLoginModule.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    public final boolean logout() throws LoginException {
+        logger.log(Level.FINEST, "Logging out " + SecurityUtil.getCredentialsFullName(credentials));
+        clearSubject();
+        return true;
+    }
+
+    /**
+     * Tidy up the Subject
+     */
+    private void clearSubject() {
```
Comment (kwart) [inline enterprise/jaas-security/src/main/java/ClientLoginModule.java]: Login modules should only clean-up their stuff (i.e. avoid removing data introduced by other login modules). It's a minor thing in this case, but if users use code samples as a quick-start for their implementations this should be done right.

File: enterprise/jaas-security/src/main/java/ClientLoginModule.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+        String username = credentials.getUsername();
+        String password = allowedUsersMap.get(username);
+        boolean loginCheckOk = false;
+
+        if (password != null){
+            if(password.equals(credentials.getPassword())){
+                String userGroup = userGroups.get(username);
+                if (userGroup != null){
+                    userGroupCredentials = new UserGroupCredentials(credentials.getEndpoint(),userGroup);
```
Comment (kwart) [inline enterprise/jaas-security/src/main/java/ClientLoginModule.java]: We don't add the authentication name part of the information into the Subject. Nonetheless, it's not a problem for this code sample.

File: enterprise/jaas-security/src/main/java/ClientLoginModule.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            logger.log(Level.WARNING, "Credentials could not be retrieved!");
+            return false;
+        }
+        logger.log(Level.INFO, "Authenticating " + SecurityUtil.getCredentialsFullName(credentials));
+
+        if (credentials instanceof UsernamePasswordCredentials){
+            loginOk = doLoginCheck((UsernamePasswordCredentials) credentials);
+        }
+
+        return loginOk;
```
[the remark this answers is not in the document]
Comment (dbrimley) [reply enterprise/jaas-security/src/main/java/ClientLoginModule.java]: OK, will change.

File: enterprise/jaas-security/src/main/java/ClientLoginModule.java (hunk 4 above)
Comment (dbrimley) [reply enterprise/jaas-security/src/main/java/ClientLoginModule.java]: I only clean up if the whole chain is aborted or if there is a logout event, so I think it's fine.

**Comment by kwart** [comment]
David, could you rebase to the current master and squash the commits?

**Comment by mmedenjak** [comment]
@kwart do you think this code sample is still useful? If so, we can rebase and squash without David's involvement.

**Comment by kwart** [comment]
We don't need to include this code sample, I think.

We have 2 other samples implementing custom login modules:
* `enterprise/client-custom-credentials`
* `enterprise/client-token-credentials`

If there is a need for more, we can always resurrect this PR.

Closing for now, feel free to reopen if necessary.

**Pull request closed without merging**